### ISSUE TITLE
Leaving username empty in login form results in user seeing a 500 error

### ISSUE THREAD

**Issue by george-hawkins-aa** [stormpath/stormpath-sdk-java#498]
If the user enters an invalid username in the login form then they see a reasonable error message and can correct their mistake.

However if they leave the username empty they see a raw 500 error with a fairly non-end user friendly message. E.g. with the custom error controller used in most of the Spring examples I see:

![image](https://cloud.githubusercontent.com/assets/12233059/13397644/2de0f99a-defb-11e5-8de8-aad8d6db0f3d.png)

If we look at the stack trace for this issue we see there seem to be different expectations at different levels:
```
Daemon Thread [http-nio-8080-exec-6]
    DefaultUsernamePasswordRequestBuilder.setUsernameOrEmail(String) line: 37    
    StormpathAuthenticationProvider.createAuthenticationRequest(Authentication) line: 425    
    StormpathAuthenticationProvider.handleUsernamePasswordAuthentication(Authentication) line: 378    
    StormpathAuthenticationProvider.authenticate(Authentication) line: 357    
    ProviderManager.authenticate(Authentication) line: 167    
    ProviderManager.authenticate(Authentication) line: 192    
    UsernamePasswordAuthenticationFilter.attemptAuthentication(HttpServletRequest, HttpServletResponse) line: 93    
    ...
```
In the `UsernamePasswordAuthenticationFilter.attemptAuthentication(...)` code we see:
```
if (username == null) {
    username = "";
}
```
I.e. it seems to be OK with the idea of the username not being present, however then in the lower down `DefaultUsernamePasswordRequestBuilder.setUsernameOrEmail(String)` call (that we see in the stack trace above) we see:
```
Assert.hasText(usernameOrEmail, "usernameOrEmail cannot be null or empty.");
```
It's this assertion that results in things blowing up and the user seeing the 500 error.

I think an empty username should be handled as nicely, from the end user's point of view, as an invalid username.

This issue is seen when using the Spring Boot setup and using version 1.0.RC8.4.

**Comment by george-hawkins-aa** [comment]
Note: if using Chrome then the default login template stops you entering an empty username on the client side:
![image](https://cloud.githubusercontent.com/assets/12233059/13398353/4af32a1e-defe-11e5-8fa5-60a651e22b0e.png)

However on other browsers, e.g. Safari, this client side checking doesn't occur and the form gets submitted in this situation. Client side checking is nice but on the whole the server shouldn't count on it occurring.

**Comment by dogeared** [comment]
Thanks for this @george-hawkins-aa. I've added this issue to the current milestone, `1.0.RC9`

**Comment by dogeared** [comment]
I also notice that going back the `1.0.RC7`, you do *not* get a 500 error. You get the friendly "Invalid login or password" message. So, we introduced this issue sometime between `1.0.RC7` and `1.0.RC8`.

**Comment by mzumbadoate** [comment]
resolves #498

**Comment by lhazlewood** [comment]
@mzumbadoate made a [comment](https://github.com/stormpath/stormpath-sdk-java/pull/507#discussion_r55074772) for this.
